src/OrleansRuntime/Messaging/IncomingMessageAcceptor.cs

```diff
@@ -235,153 +330,176 @@ private static void AcceptCallback(IAsyncResult result)
-                    return;
-                }
-                catch (Exception ex)
-                {
-                    rcc.IMA.Log.Warn(ErrorCode.Messaging_ExceptionReceiving, "Exception while completing a receive from " + rcc.Sock.RemoteEndPoint, ex);
-                    // Either there was a network error or the socket is being closed. Either way, just clean up and return.
-                    rcc.IMA.SafeCloseSocket(rcc.Sock);
-                    return;
-                }
+        private SocketAsyncEventArgs GetSocketReceiveAsyncEventArgs(Socket sock)
+        {
+            var saea = receiveEventArgsPool.Allocate();
+            var token = ((ReceiveCallbackContext) saea.SocketAsyncEventArgs.UserToken);
+            token.IMA = this;
+            token.Socket = sock;
+            checkedOutSocketEventArgsCounter.Increment();
+            return saea.SocketAsyncEventArgs;
+        }
 
-                //rcc.IMA.log.Verbose("Receive completed with " + bytes.ToString(CultureInfo.InvariantCulture) + " bytes");
-                if (bytes == 0)
-                {
-                    // Socket was closed by the sender. so close our end
-                    rcc.IMA.SafeCloseSocket(rcc.Sock);
-                    // And we're done
-                    return;
-                }
+        private static SaeaPoolWrapper CreateSocketReceiveAsyncEventArgsPoolWrapper()
+        {
+            SocketAsyncEventArgs readEventArgs = new SocketAsyncEventArgs();
+            readEventArgs.Completed += OnReceiveCompleted;
 
-                // Process the buffer we received
-                try
-                {
-                    rcc.ProcessReceivedBuffer(bytes);
-                }
-                catch (Exception ex)
-                {
-                    rcc.IMA.Log.Error(ErrorCode.Messaging_IMA_BadBufferReceived,
-                        String.Format("ProcessReceivedBuffer exception with RemoteEndPoint {0}: ",
-                            rcc.RemoteEndPoint), ex);
-                    // There was a problem with the buffer, presumably data corruption, so give up
-                    rcc.IMA.SafeCloseSocket(rcc.Sock);
-                    // And we're done
-                    return;
-                }
+            var buffer = new byte[SocketBufferSize];
+
+            // SocketAsyncEventArgs and ReceiveCallbackContext's buffer shares the same buffer list with pinned arrays.
+            readEventArgs.SetBuffer(buffer, 0, buffer.Length);
+            var poolWrapper = new SaeaPoolWrapper(readEventArgs);
+
+            // Creates with incomplete state: IMA should be set before using
+            readEventArgs.UserToken = new ReceiveCallbackContext(null, null, poolWrapper);
+            allocatedSocketEventArgsCounter.Increment();
+            return poolWrapper;
+        }
+
+        private void FreeSocketAsyncEventArgs(SocketAsyncEventArgs args)
+        {
+            var receiveToken = (ReceiveCallbackContext) args.UserToken;
```

Comment:
If we're going to reuse the ReceiveCallbackContext, we need to make sure to reset it's buffers.  Any errors that could change the socket feeding a receiver buffer must reset the buffer.